src/messages/MClientReply.h

```diff
@@ -391,27 +388,27 @@ class MClientReply final : public MMDSOp {
     using ceph::decode;
     auto p = payload.cbegin();
     decode(head, p);
+    // for the backward compatability, assuming LINUX error codes on the wire,
+    // errorcode32_t implements conversion from/to different host error codes
+    if (header.version >= HEAD_VERSION) {
+      errorcode32_t temp {static_cast<__s32>(head.result)};
+      temp.convert_decode();
+      head.result = static_cast<__u32>(temp.code);
+    }
     decode(trace_bl, p);
     decode(extra_bl, p);
     decode(snapbl, p);
-    // for the backward compatability, assuming LINUX error codes on the wire,
-    // we use the head.result as a value for error_code BEFORE decoding error_code.
-    // for the ver2 messages, the result is set in the error_code.result by sender
-    if (header.version < HEAD_VERSION) {
-      error_code.code = error_code.convert_decode(head.result);
-    } else {
-      decode(error_code, p);
-    }
     ceph_assert(p.end());
   }
   void encode_payload(uint64_t features) override {
     using ceph::encode;
-    head.result = error_code.convert_encode();
+    // errorcode32_t implements conversion from/to different host error codes
+    errorcode32_t temp {static_cast<__s32>(head.result)};
+    head.result = static_cast<__u32>(temp.convert_encode());
```

Comment:
ditto, we're first casting the `head.result` to 32 bit signed and then to 32 bit unsigned?

Reply:
see comments above